trio/_socket.py

```diff
@@ -520,16 +528,211 @@ async def _resolve_address_nocp(
     return normed
 
 
-# TODO: stopping users from initializing this type should be done in a different way,
-# so SocketType can be used as a type. Note that this is *far* from trivial without
-# breaking subclasses of SocketType. Can maybe add abstract methods to SocketType,
-# or rename _SocketType.
 class SocketType:
-    def __init__(self) -> NoReturn:
-        raise TypeError(
-            "SocketType is an abstract class; use trio.socket.socket if you "
-            "want to construct a socket object"
-        )
+    def __init__(self) -> None:
+        if type(self) == SocketType:
+            raise TypeError(
+                "SocketType is an abstract class; use trio.socket.socket if you "
+                "want to construct a socket object"
+            )
+
+    if TYPE_CHECKING:
+
+        def detach(self: SocketType) -> int:
+            ...
+
+        def fileno(self: SocketType) -> int:
+            ...
+
+        def getpeername(self: SocketType) -> Any:
+            ...
+
+        def getsockname(self: SocketType) -> Any:
```

Comment:
Not type checking that return type is instance of `object` or does this lead to issues?